swaybar/bar.c

```diff
@@ -178,23 +233,54 @@ void bar_setup(struct bar *bar, const char *socket_path, const char *bar_id) {
 	}
 	/* spawn status command */
 	spawn_status_cmd_proc(bar);
+
+#ifdef ENABLE_TRAY
+	// We should have at least one output to serve the tray to
+	if (!swaybar.config->tray_output || strcmp(swaybar.config->tray_output, "none") != 0) {
+		/* Connect to the D-Bus */
+		dbus_init();
+
+		/* Start the SNI watcher */
+		init_sni_watcher();
+
+		/* Start the SNI host */
+		init_tray();
+
+		/* Start xembedsniproxy */
+		spawn_xembed_sni_proxy();
+	}
+#endif
```

Comment:
This one too, just make it a simple

```c
#ifdef ENABLE_TRAY
    init_tray();
#endif
```

Then make init_tray do all this work. Try to limit the tray code in this file to simple calls to another function.